python_modules/libraries/dagster-k8s/dagster_k8s/external_resource.py

```diff
@@ -0,0 +1,204 @@
+        ):
+            namespace = namespace or "default"
+            pod_name = get_pod_name(context.run_id, context.op.name)
+            pod_body = build_pod_body(
+                pod_name=pod_name,
+                image=image,
+                command=command,
+                env_vars={
+                    **self.get_base_env(),
+                    **protocol_env_vars,
+                    **(self.env or {}),
+                    **(env or {}),
+                },
+                base_pod_meta=base_pod_meta,
+                base_pod_spec=base_pod_spec,
+            )
+            client.core_api.create_namespaced_pod(namespace, pod_body)
+            try:
+                # if were doing direct pod reading, wait for pod to start and then stream logs out
+                if isinstance(resolved_msg_reader, K8sPodLogsMessageReader):
+                    client.wait_for_pod(
+                        pod_name,
+                        namespace,
+                        wait_for_state=WaitForPodState.Ready,
+                    )
+                    resolved_msg_reader.consume_pod_logs(
+                        ext_context=ext_context,
+                        client=client,
+                        pod_name=pod_name,
+                        namespace=namespace,
+                    )
+                else:
+                    # if were not doing direct log reading, just wait for pod to finish
+                    client.wait_for_pod(
+                        pod_name,
+                        namespace,
+                        wait_for_state=WaitForPodState.Terminated,
+                    )
+            finally:
+                client.core_api.delete_namespaced_pod(pod_name, namespace)
+
+
+def build_pod_body(
+    pod_name: str,
+    image: Optional[str],
+    command: Optional[Union[str, Sequence[str]]],
+    env_vars: Mapping[str, str],
+    base_pod_meta: Optional[Mapping[str, Any]],
+    base_pod_spec: Optional[Mapping[str, Any]],
+):
+    meta = {
+        **(base_pod_meta or {}),
+        "name": pod_name,
+    }
+
+    spec = {**(base_pod_spec or {})}
+    if "containers" not in spec:
+        spec["containers"] = [{}]
+
+    if "image" not in spec["containers"][0] and not image:
+        raise DagsterInvariantViolationError(
+            "Must specify image property or provide base_pod_spec with one set."
+        )
+
+    if "name" not in spec["containers"][0]:
+        spec["containers"][0]["name"] = DEFAULT_CONTAINER_NAME
+
+    if image:
+        spec["containers"][0]["image"] = image
+
+    if command:
+        spec["containers"][0]["command"] = command
+
+    if "env" not in spec["containers"][0]:
+        spec["containers"][0]["env"] = []
+
+    spec["containers"][0]["env"].extend({"name": k, "value": v} for k, v in env_vars.items())
+
+    return k8s_model_from_dict(
+        kubernetes.client.V1Pod,
+        {
+            "metadata": meta,
+            "spec": spec,
+        },
+    )
```

Comment:
🤔 this is pretty gnarly. joys of loosely typed apis

Reply:
the k8s python api objects are mutable (ie `kubernetes.client.V1Pod`) so i started trying those, but they do validation on construction so you need to get things filled out before you use them - which led me to the pattern we take elsewhere in the code base of doing a bunch of `dict` shenanigans and then one final construction (with its validation)  at the end. Bonus going through `k8s_model_from_dict` supports flexibility in `camelCase` or `snake_case` for all the props

Reply:
Makes sense. Couldn't let that pass without saying _something_.